grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/GrpcDocServicePlugin.java

```diff
@@ -143,75 +147,69 @@ public ServiceSpecification generateSpecification(Set<ServiceConfig> serviceConf
         requireNonNull(serviceConfigs, "serviceConfigs");
         requireNonNull(filter, "filter");
 
+        final Set<GrpcService> addedService = new HashSet<>();
         final ImmutableList.Builder<HttpEndpoint> httpEndpoints = ImmutableList.builder();
         final ServiceInfosBuilder serviceInfosBuilder = new ServiceInfosBuilder();
         for (ServiceConfig serviceConfig : serviceConfigs) {
             final GrpcService grpcService = serviceConfig.service().as(GrpcService.class);
             assert grpcService != null;
 
+            if (addedService.add(grpcService)) {
+                addServiceDescriptor(serviceInfosBuilder, grpcService);
+            }
+
             if (grpcService instanceof HttpEndpointSupport) {
+                // grpcService is a HttpJsonTranscodingService.
                 final HttpEndpointSpecification spec =
-                        ((HttpEndpointSupport) grpcService).httpEndpointSpecification(serviceConfig.route());
-                if (spec != null && filter.test(NAME, spec.serviceName(), spec.methodName())) {
-                    httpEndpoints.add(new HttpEndpoint(serviceConfig, spec));
+                        ((HttpEndpointSupport) grpcService).httpEndpointSpecification(
+                                serviceConfig.mappedRoute()); // Use mappedRoute to find the specification.
+                if (spec != null) {
+                    if (filter.test(NAME, spec.serviceName(), spec.methodName())) {
+                        httpEndpoints.add(new HttpEndpoint(serviceConfig,
+                                                           // Use route which has the full path.
+                                                           spec.withRoute(serviceConfig.route())));
+                    }
+                    continue;
```

Comment:
I'm failing to understand why we `continue` if `spec` isn't `null` - would you mind explaining? 😅

Reply:
Because we added this route using `HttpEndpointSpecification`, we don't have to add it again.
I left a comment below `if (grpcService instanceof HttpEndpointSupport) {`. Please let me know if it's unclear. 😉

Reply:
Ah got it 👍  Thanks!